File: kmp-workshop/iosApp/iosApp/BookService.swift
//
// Created by Dominik Chmelík on 14.10.2025.
//

import Foundation

final class BookService {
    static let instance = BookService()

    func fetchBooks() -> [Book] {
        return [
            Book(id: "1", name: "The Hobbit", pageCount: 300, rating: 4.5),
            Book(id: "2", name: "The Lord of the Rings", pageCount: 500, rating: 4.2),
            Book(id: "3", name: "1984", pageCount: 400, rating: 4.7),
            Book(id: "4", name: "To Kill a Mockingbird", pageCount: 600, rating: 4.5),
            Book(id: "5", name: "The War of the Worlds", pageCount: 700, rating: 3.9),
            Book(id: "6", name: "Pride and Prejudice", pageCount: 800, rating: 4.5),
            Book(id: "7", name: "Moby-Dick", pageCount: 900, rating: 4.3),
            Book(id: "8", name: "The Catcher in the Rye", pageCount: 1000, rating: 4.5),
            Book(id: "9", name: "The Alchemist", pageCount: 1200, rating: 4.5),
            Book(id: "10", name: "The Lord of the Flies", pageCount: 1100, rating: 4.0)
        ]
    }

    func isBookLiked(id: String) -> Bool {
        return UserDefaults.standard.bool(forKey: "book-liked-\(id)")
    }

    func likeBook(id: String) {
        UserDefaults.standard.set(true, forKey: "book-liked-\(id)")
    }

    func unlikeBook(id: String) {
        UserDefaults.standard.set(false, forKey: "book-liked-\(id)")
    }
}

File: kmp-workshop/shared/src/commonMain/kotlin/com/chmelik/kmp_workshop/BookService.kt
package com.chmelik.kmp_workshop

class BookService(
    private val storage: Storage
) {
    fun fetchBooks(): List<Book> = listOf(
        Book("1", "The Hobbit", 300, 4.5),
        Book("2", "The Lord of the Rings", 500, 4.2),
        Book("3", "1984", 400, 4.7),
        Book("4", "To Kill a Mockingbird", 600, 4.5),
        Book("5", "The War of the Worlds", 700, 3.9),
        Book("6", "Pride and Prejudice", 800, 4.5),
        Book("7", "Moby-Dick", 900, 4.3),
        Book("8", "The Catcher in the Rye", 1000, 4.5),
        Book("9", "The Alchemist", 1200, 4.5),
        Book("10", "The Lord of the Flies", 1100, 4.0)
    )

    fun isBookLiked(id: String): Boolean {
        return storage.getBool("book-liked-$id")
    }

    fun likeBook(id: String) {
        storage.setBool(true, "book-liked-$id")
    }

    fun unlikeBook(id: String) {
        storage.setBool(false, "book-liked-$id")
    }
}
